substratevm/src/com.oracle.svm.native.libchelper/src/cpuid.c

```diff
@@ -75,106 +94,509 @@ int get_cpuid (unsigned int leaf, unsigned int *eax, unsigned int *ebx, unsigned
-    get_cpuid_count (7, 0, &eax, &ebx, &ecx, &edx);
-
-    features->fERMS     = !!(ebx & bit_ERMS_compat);
-    features->fAVX2     = !!(ebx & bit_AVX2_compat);
-    features->fBMI1     = !!(ebx & bit_BMI1_compat);
-    features->fBMI2     = !!(ebx & bit_BMI2_compat);
-    features->fRTM      = !!(ebx & bit_RTM_compat);
-    features->fADX      = !!(ebx & bit_ADX_compat);
-    features->fAVX512F  = !!(ebx & bit_AVX512F_compat);
-    features->fAVX512DQ = !!(ebx & bit_AVX512DQ_compat);
-    features->fAVX512PF = !!(ebx & bit_AVX512PF_compat);
-    features->fAVX512ER = !!(ebx & bit_AVX512ER_compat);
-    features->fAVX512CD = !!(ebx & bit_AVX512CD_compat);
-    features->fAVX512BW = !!(ebx & bit_AVX512BW_compat);
-    features->fAVX512VL = !!(ebx & bit_AVX512VL_compat);
-    features->fSHA      = !!(ebx & bit_SHA_compat);
+
+static char supports_processor_topology() {
+  return (_cpuid_info.std_max_function >= 0xB) &&
+          // eax[4:0] | ebx[0:15] == 0 indicates invalid topology level.
+          // Some cpus have max cpuid >= 0xB but do not support processor topology.
+          (((_cpuid_info.tpl_cpuidB0_eax & 0x1f) | _cpuid_info.tpl_cpuidB0_ebx.bits.logical_cpus) != 0);
+}
+
+static uint32_t cores_per_cpu() {
+  uint32_t result = 1;
+  if (is_intel())
+  {
+    char supports_topology = supports_processor_topology();
+    if (supports_topology)
+    {
+      result = _cpuid_info.tpl_cpuidB1_ebx.bits.logical_cpus /
+               _cpuid_info.tpl_cpuidB0_ebx.bits.logical_cpus;
+    }
+    if (!supports_topology || result == 0)
+    {
+      result = (_cpuid_info.dcp_cpuid4_eax.bits.cores_per_cpu + 1);
+    }
   }
+  else if (is_amd_family())
+  {
+    result = (_cpuid_info.ext_cpuid8_ecx.bits.cores_per_cpu + 1);
+  }
+  else if (is_zx())
+  {
+    char supports_topology = supports_processor_topology();
+    if (supports_topology)
+    {
+      result = _cpuid_info.tpl_cpuidB1_ebx.bits.logical_cpus /
+               _cpuid_info.tpl_cpuidB0_ebx.bits.logical_cpus;
+    }
+    if (!supports_topology || result == 0)
+    {
+      result = (_cpuid_info.dcp_cpuid4_eax.bits.cores_per_cpu + 1);
+    }
+  }
+  return result;
+}
 
-  // figuring out extended features
-  get_cpuid (0x80000000, &ext_level, &ebx, &ecx, &edx);
+static uint32_t threads_per_core() {
+  uint32_t result = 1;
+  if (is_intel() && supports_processor_topology())
+  {
+    result = _cpuid_info.tpl_cpuidB0_ebx.bits.logical_cpus;
+  }
+  else if (is_zx() && supports_processor_topology())
+  {
+    result = _cpuid_info.tpl_cpuidB0_ebx.bits.logical_cpus;
+  }
+  else if (_cpuid_info.std_cpuid1_edx.bits.ht != 0)
+  {
+    if (cpu_family() >= 0x17)
+    {
+      result = _cpuid_info.ext_cpuid1E_ebx.bits.threads_per_core + 1;
+    }
+    else
+    {
+      result = _cpuid_info.std_cpuid1_ebx.bits.threads_per_cpu /
+               cores_per_cpu();
+    }
+  }
+  return (result == 0 ? 1 : result);
+}
+
+// This is ported from the HotSpot stub get_cpu_info_stub within vm_version_x86.cpp
```

Comment:
@mur47x111 recently added support for annotating Java code ported from HotSpot: 9c2cd4e855ae01b766597d69f26b5af6d6a8d729
Maybe you could use that to ensure the Graal daily CI builders notify us when the relevant HotSpot sources change.

Reply:
This can be done (if at all) in a follow up PR.